### PR TITLE
Pages 3.0 To Use Schema Resources From Servlet 5.0 API

### PR DESCRIPTION
**Pull request by volosied**
Potential Fix for #12637

-  Deleted the io.openliberty.jakarta.pages.30 resources package (jakarta/servlet/jsp/resources/*) 
-  jspxml.dtd, jspxml.xsd are also deleted as they are not needed for Pages 3.0. 
- The change does not affect the behavior the jsp features, but only uses the newer resources files for `pages-3.0`. 
- Transformation rules were updated to retrieve resources from the servlet 5.0 api bundle rather than the pages 3.0 bundle.

Only issue is that the other JSP bundles need this package  above for bundle resolution.  Since we delete j`akarta/servlet/jsp/resources/` then the transformed JSP bundles (i.e com.ibm.ws.jsp) will not be able to find `jakarta/servlet/jsp/resources`.  There are no resolution problems with JSP 2.2 or 2.3 as this `javax/servlet/jsp/resources` still exists in the older API bundles. 

Example error message:
```
[ERROR   ] CWWKE0702E: Could not resolve module: com.ibm.ws.jsp.2.3.jakarta [81]
  Unresolved requirement: Import-Package: com.ibm.ws.jsp.translator; version="2.2.0"
    -> Export-Package: com.ibm.ws.jsp.translator; bundle-symbolic-name="com.ibm.ws.jsp.jakarta"; bundle-version="1.0.55.202107292055"; version="2.2.0"; uses:="com.ibm.websphere.ras.annotation,com.ibm.ws.jsp,com.ibm.ws.jsp.configuration,com.ibm.ws.jsp.translator.utils,org.w3c.dom"
       com.ibm.ws.jsp.jakarta [82]
         Unresolved requirement: Import-Package: junit.framework; resolution:="optional"
         Unresolved requirement: Import-Package: junit.textui; resolution:="optional"
         Unresolved requirement: Import-Package: jakarta.servlet.jsp.resources; version="[3.0.0,4.0.0)"

```

The only workaround I've been able to find is to continue to export this`jakarta/servlet/jsp/resources` package even though there aren't any files (class or resource) located within in.  No files are used from this package in the other bundle (nor is is anything available to any applications), so we shouldn't encounter any missing class/resource exceptions.

### CODE REVIEW

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java [hunk 1]
```diff
@@ -598,7 +608,12 @@ public static synchronized GlobalTagLibraryCache getGlobalTagLibraryCache() {
      */
     @Reference(cardinality=ReferenceCardinality.MULTIPLE, policy=ReferencePolicy.DYNAMIC)
     protected void setGlobalTagLibConfig(GlobalTagLibConfig globalTagLibConfig) {
-            getGlobalTagLibraryCache().addGlobalTagLibConfig(globalTagLibConfig);
+            // Wait for pages version to be loaded first before calling getGlobalTagLibraryCache
+            if(!isPagesVersionLoaded){
+                storedGlobalTagLibConfigs.add(globalTagLibConfig);
+            } else {
+                getGlobalTagLibraryCache().addGlobalTagLibConfig(globalTagLibConfig);
+            }
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java]: _setVersion_ is called after _setGlobalTagLibConfig_, so we just store the globalTagLibConfig and then add it to the GlobalTagLibraryCache in the setVersion

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.jsp_fat,com.ibm.ws.jsp.2.3_fat, com.ibm.ws.jsf.2.2_fat, com.ibm.ws.jsf.2.3_fat

**Comment by volosied** [comment]
#build

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/TldParser.java [hunk 2]
```diff
@@ -82,6 +83,17 @@
...
+    /* Servlet 5.0 API contains these JSP files */
+
+    public static final String EE9_TAGLIB_DTD_PUBLIC_ID_11 = "-//Sun Microsystems, Inc.//DTD JSP Tag Library 1.1//EN";
+    public static final String EE9_TAGLIB_DTD_RESOURCE_PATH_11 = "/jakarta/servlet/resources/web-jsptaglibrary_1_1.dtd";
+    
+    public static final String EE9_TAGLIB_DTD_PUBLIC_ID_12 = "-//Sun Microsystems, Inc.//DTD JSP Tag Library 1.2//EN";
+    public static final String EE9_TAGLIB_DTD_RESOURCE_PATH_12 = "/jakarta/servlet/resources/web-jsptaglibrary_1_2.dtd";
+
+    public static final String EE9_TAGLIB_XSD_SYSTEM_ID_20 = "web-jsptaglibrary_2_0.xsd";
+    public static final String EE9_TAGLIB_XSD_RESOURCE_PATH_20 = "/jakarta/servlet/resources/web-jsptaglibrary_2_0.xsd";
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/TldParser.java]: The following don't change, so we shouldn't duplicate these values. Only make add variables for the new paths. 

- EE9_TAGLIB_DTD_PUBLIC_ID_11
- EE9_TAGLIB_DTD_PUBLIC_ID_12
- EE9_TAGLIB_XSD_SYSTEM_ID_20

File: dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java [hunk 3]
```diff
@@ -641,5 +656,13 @@ public static String getLoadedPagesSpecLevel() {
...
         return JSPExtensionFactory.loadedSpecLevel;
     }
+    
+    public static boolean isPages30orHigher(){
+        String version = getLoadedPagesSpecLevel();
+        if(version.equals("2.2") || version.equals("2.3")){
+            return false;
+        }
+        return true;
+    }
```
Comment (volosied) [inline dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/webcontainerext/JSPExtensionFactory.java:666]: Helper method since check for page-3.0 or higher very often.

**Review by epj — APPROVED** [review]
looks good

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat

**Comment by volosied** [comment]
#build

File: dev/io.openliberty.jakarta.pages.3.0/bnd.bnd [hunk 4]
```diff
@@ -14,13 +14,15 @@ bVersion=1.0
...
 Bundle-Description: Jakarta Server Pages, version 3.0
 
+# Note: No files under jakarta.servlet.jsp.resources, but it's needed to resolve other jsp bundles (i.e. com.ibm.jsp.2.3.jakarta)
 Export-Package: \
 	jakarta.servlet.jsp; version="3.0.0",\
 	jakarta.servlet.jsp.el; version="3.0.0",\
-	jakarta.servlet.jsp.resources; version="3.0.0",\
-	jakarta.servlet.jsp.tagext; version="3.0.0"
+	jakarta.servlet.jsp.tagext; version="3.0.0",\
+	jakarta.servlet.jsp.resources; version="3.0.0"
```
Comment (pnicolucci) [inline dev/io.openliberty.jakarta.pages.3.0/bnd.bnd]: any reason the order was changed here?

File: dev/wlp-jakartaee-transform/rules/jakarta-direct.properties [hunk 5]
```diff
@@ -64,9 +64,9 @@ javax.servlet.jsp.jstl.sql.userName=jakarta.servlet.jsp.jstl.sql.userName
 javax.faces.validator.beanValidator.ValidatorFactory=jakarta.faces.validator.beanValidator.ValidatorFactory
 
 # Jakarta Server Pages TLD Locations
-/javax/servlet/jsp/resources/web-jsptaglibrary_1_1.dtd=/jakarta/servlet/jsp/resources/web-jsptaglibrary_1_1.dtd
-/javax/servlet/jsp/resources/web-jsptaglibrary_1_2.dtd=/jakarta/servlet/jsp/resources/web-jsptaglibrary_1_2.dtd
-/javax/servlet/jsp/resources/web-jsptaglibrary_2_0.xsd=/jakarta/servlet/jsp/resources/web-jsptaglibrary_2_0.xsd
+/javax/servlet/jsp/resources/web-jsptaglibrary_1_1.dtd=/jakarta/servlet/resources/web-jsptaglibrary_1_1.dtd
+/javax/servlet/jsp/resources/web-jsptaglibrary_1_2.dtd=/jakarta/servlet/resources/web-jsptaglibrary_1_2.dtd
+/javax/servlet/jsp/resources/web-jsptaglibrary_2_0.xsd=/jakarta/servlet/resources/web-jsptaglibrary_2_0.xsd
```
Comment (pnicolucci) [inline dev/wlp-jakartaee-transform/rules/jakarta-direct.properties:69]: Just to ensure I understand the pages resources now reside within the servlet bundle in the  `jakarta.servlet.resources` package?

File: dev/io.openliberty.jakarta.pages.3.0/bnd.bnd (hunk 4 above)
Comment (volosied) [reply dev/io.openliberty.jakarta.pages.3.0/bnd.bnd]: Not intentional. I had added a comment between the exports but then opted to place it above on line 17 instead.

File: dev/wlp-jakartaee-transform/rules/jakarta-direct.properties (hunk 5 above)
Comment (volosied) [reply dev/wlp-jakartaee-transform/rules/jakarta-direct.properties:69]: Correct

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
One more build for safe measure.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsp.2.3_fat

File: dev/io.openliberty.jakarta.pages.3.0/bnd.bnd (hunk 4 above)
Comment (pnicolucci) [reply dev/io.openliberty.jakarta.pages.3.0/bnd.bnd]: Can we change it back? It makes looking for a package easier If they are in alphabetical order.

**Review by pnicolucci — COMMENTED** [review]
I'd really just like to see the reordering of the packages in the bnd.bnd removed to keep them in alphabetical order.

File: dev/io.openliberty.jakarta.pages.3.0/bnd.bnd (hunk 4 above)
Comment (volosied) [reply dev/io.openliberty.jakarta.pages.3.0/bnd.bnd]: Done! Changed to the original order.

**Review by pnicolucci — APPROVED** [review]

**Comment by pnicolucci** [comment]
@volosied  I approved these changes please just ensure that your latest personal build is good before merging.

**Comment by pnicolucci** [comment]
#run-libby-bot

**Review by pnicolucci — COMMENTED** [review]
Sorry one more comment, can you update the copyright on the `io.openliberty.jakarta.pages.3.0/bnd.bnd`?

**Comment by volosied** [comment]
JSP tests all passed, but the build had a lot of unrelated errors?  I rebased to to submit one more buid.

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#run-libby-bot

**Pull request merged**